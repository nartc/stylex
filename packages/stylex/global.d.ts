/**
 * Copyright (c) Ladifire, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as CSS from 'csstype';

export interface CSSProperties extends CSS.Properties<string | number> {
  /**
   * The index signature was removed to enable closed typing for style
   * using CSSType. You're able to use type assertion or module augmentation
   * to add properties or an index signature of your own.
   *
   * For examples and more information, visit:
   * https://github.com/frenic/csstype#what-should-i-do-when-i-get-type-errors
   */
}

export type AdvancedPseudos =
  | ":-moz-any()"
  | ":-moz-dir"
  | ":-webkit-any()"
  | "::cue"
  | "::cue-region"
  | "::part"
  | "::slotted"
  | ":dir"
  | ":has"
  | ":host"
  | ":host-context"
  | ":is"
  | ":lang"
  | ":matches()"
  | ":not"
  | ":nth-child"
  | ":nth-last-child"
  | ":nth-last-of-type"
  | ":nth-of-type"
  | ":where";

export type SimplePseudos =
  | ":-khtml-any-link"
  | ":-moz-any-link"
  | ":-moz-focusring"
  | ":-moz-full-screen"
  | ":-moz-placeholder"
  | ":-moz-read-only"
  | ":-moz-read-write"
  | ":-ms-fullscreen"
  | ":-ms-input-placeholder"
  | ":-webkit-any-link"
  | ":-webkit-full-screen"
  | "::-moz-placeholder"
  | "::-moz-progress-bar"
  | "::-moz-range-progress"
  | "::-moz-range-thumb"
  | "::-moz-range-track"
  | "::-moz-selection"
  | "::-ms-backdrop"
  | "::-ms-browse"
  | "::-ms-check"
  | "::-ms-clear"
  | "::-ms-fill"
  | "::-ms-fill-lower"
  | "::-ms-fill-upper"
  | "::-ms-input-placeholder"
  | "::-ms-reveal"
  | "::-ms-thumb"
  | "::-ms-ticks-after"
  | "::-ms-ticks-before"
  | "::-ms-tooltip"
  | "::-ms-track"
  | "::-ms-value"
  | "::-webkit-backdrop"
  | "::-webkit-input-placeholder"
  | "::-webkit-progress-bar"
  | "::-webkit-progress-inner-value"
  | "::-webkit-progress-value"
  | "::-webkit-slider-runnable-track"
  | "::-webkit-slider-thumb"
  | "::after"
  | "::backdrop"
  | "::before"
  | "::cue"
  | "::cue-region"
  | "::first-letter"
  | "::first-line"
  | "::grammar-error"
  | "::marker"
  | "::placeholder"
  | "::selection"
  | "::spelling-error"
  | ":active"
  | ":after"
  | ":any-link"
  | ":before"
  | ":blank"
  | ":checked"
  | ":default"
  | ":defined"
  | ":disabled"
  | ":empty"
  | ":enabled"
  | ":first"
  | ":first-child"
  | ":first-letter"
  | ":first-line"
  | ":first-of-type"
  | ":focus"
  | ":focus-visible"
  | ":focus-within"
  | ":fullscreen"
  | ":hover"
  | ":in-range"
  | ":indeterminate"
  | ":invalid"
  | ":last-child"
  | ":last-of-type"
  | ":left"
  | ":link"
  | ":only-child"
  | ":only-of-type"
  | ":optional"
  | ":out-of-range"
  | ":placeholder-shown"
  | ":read-only"
  | ":read-write"
  | ":required"
  | ":right"
  | ":root"
  | ":scope"
  | ":target"
  | ":valid"
  | ":visited";

export type Pseudos = AdvancedPseudos | SimplePseudos;

// TODO: missing media query
export type CSSPropertiesWithNestedPseudo<T extends string> = {
  [key in T]: CSSProperties
} | {
  [key in T]: {
    [key in Pseudos]?: CSSProperties
  }
}

export type CSSObject = CSSProperties & {
  [key in Pseudos]?: CSSProperties
}

