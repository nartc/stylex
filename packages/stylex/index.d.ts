/**
 * Copyright (c) Ladifire, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// TODO: Phần mô tả ở dưới đây mình tạm để tiếng việt cho một số bạn không rành TA tham gia
// TODO: Xem ví dụ thực tế khi sử dụng trong file: ./packages/uses_cases_DEPRECATED/uses_cases_DEPRECATED.tsx
// TODO: Hàm create dưới đây là đoạn mình viết, nhưng nó vẫn chưa thực sự chuẩn với mọi trường hợp, nên cũng cần review lại
// TODO: (Khi viết type definition nhớ để ý pseudo, media query nhé)

import {CSSPropertiesWithNestedPseudo, CSSObject, CSSProperties} from './global';

// tslint:disable-next-line:export-just-namespace
export = stylex;
export as namespace stylex;

declare function stylex<T extends string>(...style: (CSSObject | CSSObject[])[]): string;

declare namespace stylex {
  // TODO 1: phần này chưa có definition
  // vd 1: className={stylex(styles.root)} <== truyền 1 object stylex
  // vd 2: className={stylex([styles.root, styles.container, ....])} <== truyền 1 mảng các objects stylex

  // TODO 2: stylex.create(...)
  // vd:
  // const styles = stylex.create({
  //   root: {
  //     alignItems: 'center',
  //     cursor: 'pointer',
  //     display: 'flex',
  //     marginTop: 0,
  //     marginRight: 12,
  //     marginBottom: 0,
  //     marginLeft: 8,
  //   },
  // });
  function create<T extends string | number>(styles: CSSPropertiesWithNestedPseudo<T>):
    { [key in T]: CSSProperties } & (
    (...styles: (
      | keyof T
      | boolean
      | undefined
      | null
      | { [key in keyof T]?: boolean | undefined | null }
      )[]) => string
    )

  // TODO 3: stylex.dedupe(...)
  // Hàm này dedupe (có thể hiểu là sẽ lấy các giá trị sau đè lên các giá trị trước) các dối tượng theo một điều kiện nào đó
  // kết quả trả về là một chuỗi string gồm các className ("aadsdfs etrewtre asdfs454s wer1werterw5 ....")
  // Cấu trúc: vd1
  //     stylex.dedupe({
  //       display: "flex",
  //       flexDirection: "row",
  //       flexWrap: "wrap"
  //     }, d === "narrow" ? {
  //       margin: -2,
  //     } : null, d === "wide" ? {
  //       margin: -3,
  //     } : null)
  //
  // ví dụ khác:
  //
  // <div
  //   className={stylex.dedupe(
  //     {
  //       position: 'absolute',
  //       top: 0,
  //       left: 0,
  //       bottom: 0,
  //       right: 0,
  //       transitionTimingFunction: 'var(--lf-animation-fade-out)',
  //       pointerEvents: 'none',
  //       transitionDuration: 'var(--lf-duration-extra-extra-short-out)',
  //       transitionProperty: 'opacity',
  //       cursor: 'pointer',
  //     },
  //     !isLink ? {
  //       backgroundColor: 'var(--hover-overlay)',
  //     } : null,
  //     !isLink && (isHovered || selected) ? {
  //       opacity: 1,
  //     } : {
  //       opacity: 0,
  //     },
  //     !isLink && isPressed ? {
  //       backgroundColor: 'var(--press-overlay)',
  //     } : null,
  //     !isLink && isPressed && variant === 'primary-deemphasized' ? {
  //       backgroundColor: 'var(--primary-deemphasized-button-pressed-overlay)',
  //     } : null,
  //     rounded ? {
  //       borderRadius: '50%',
  //     } : {
  //       borderRadius: 6,
  //     },
  //   )}
  // />
  function dedupe(...styles: CssStyle[]): string;

  // TODO 4: stylex.compose(...)
  // Hàm này merge các đối tượng stylex lại với nhau (kết quả là một đối tượng stylex duy nhất)
  // Hàm này tương tự như Object.assign
  // <div
  //   className={stylex.compose(
  //     {
  //      //... stylex object 1
  //     },
  //     {
  //       //... stylex object 1
  //     }
  //   )};
  // />
  function compose(...styles: CssStyle[]): CssStyle;

  // TODO 5: stylex.keyframes(...)
  // Hàm này khai báo một keyframes animation name
  // Ví dụ:
  // let j = stylex.create({
  //   root: {
  //     animationDirection: "alternate",
  //     animationDuration: "1s",
  //     animationIterationCount: "infinite",
  //     animationName: stylex.keyframes({
  //       '0%': {
  //         opacity: 0.25,
  //       },
  //       '100%': {
  //         opacity: 1,
  //       },
  //     }),
  //     animationTimingFunction: "steps(10,end)",
  //     backgroundColor: "var(--wash)",
  //     opacity: 0.25
  //   }
  // });
  function keyframes(rules: CssStyle): string;

  // TODO 6: stylex.inject(...)
  // inject một chuỗi css vào compiled js
  // vd: inject(".a55dalm2{background-color:red}")
  function inject(cssString: string): void;
}
