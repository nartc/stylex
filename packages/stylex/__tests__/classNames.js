/* eslint-env jest */
const babelCliTransform = require("./babel-cli-transforms");

it('string', () => {
  const input = `
    import stylex from '@ladifire-opensource/stylex';
    const styles = stylex.create({
      root: {
        color: "red",
        backgroundColor: "blue",
      },
      button: {
        color: "var(--primary-text)"
      }
    });
        var a = styles('root', 'button');
  `;
  const {code} = babelCliTransform(input);
  expect(code).toMatchSnapshot();
});
