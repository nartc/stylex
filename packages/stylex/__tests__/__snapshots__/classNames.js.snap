// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string 1`] = `
"\\"use strict\\";

var _stylex = _interopRequireDefault(require(\\"@ladifire-opensource/stylex\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

var styles = {
  root: {
    color: \\"aaaeyvqo\\",
    backgroundColor: \\"jab5yxxb\\"
  },
  button: {
    color: \\"jaarygkp\\"
  }
};
styles('root', 'button');"
`;
